fastlane/lib/fastlane/helper/adb_helper.rb

```diff
@@ -30,23 +30,74 @@ def trigger(command: nil, serial: nil)
         Action.sh(command)
       end
 
-      def device_avalaible?(serial)
-        load_all_devices
+      def device_avalaible?(serial, load_names = false)
+        load_all_devices(load_names)
         return devices.map(&:serial).include?(serial)
       end
 
-      def load_all_devices
+      def load_all_devices(load_names = false)
         self.devices = []
 
-        command = [adb_path, "devices"].join(" ")
+        command = [adb_path, "devices -l"].join(" ")
         output = Actions.sh(command, log: false)
         output.split("\n").each do |line|
-          if (result = line.match(/(.*)\tdevice$/))
+          if (result = line.match(/^(\S+)(\s+)(device )/))
             self.devices << AdbDevice.new(serial: result[1])
           end
         end
+
+        # Checks if telnet is intalled
+        # Display message to user to install it to get full value
+        has_telnet_installed = has_brew_tool!('telnet')
+
+        # Iterates devices to load name from telnet for emulators
+        # and adb from actual devices
+        if load_names
+          self.devices.each do |device|
+            if device.serial.include?('emulator-')
+              load_name_from_telnet(device) if has_telnet_installed
+            else
+              load_name_from_adb(device)
+            end
+          end
+        end
+
         self.devices
       end
+
+      private
+
+      def load_name_from_telnet(device)
+        # Gets port to connect to via telnet
+        port = device.serial.gsub('emulator-', '')
+
+        # Pipes `avd name` to telnet, sleeps to get output, and exits
+        # Runs empty echo after telnet to make Actions.sh happy
+        output = Actions.sh(
+          "sleep1; (echo 'avd name'; sleep 1; echo 'exit 0') | telnet localhost #{port}; echo ''",
+          log: false
+        )
+
+        # Gets emulator name from between the "OK"s
+        regex = /\sOK\s(.*?)\sOK\b/m
+        device.name = output.slice(regex, 1)
+      end
+
+      def load_name_from_adb(device)
+        device.name = trigger(command: "shell getprop ro.product.model", serial: device.serial).strip
+      end
+
+      def has_brew_tool!(tool)
```

Comment:
This is Mac only. Can we somehow change this to be platform independent?

Reply:
Instructions to install telnet on Windows:
https://support.code42.com/CrashPlan/6/Troubleshooting/Test_your_network_connection#Install_Telnet_on_Windows

Now we also need a way to properly find out in ruby if telnet is installed, because `which telnet` obv doesn't work on Windows.

Reply:
On Linux telnet should always be available?